pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,172 @@
+
+        # S = { c ∩ c': c, c' ∈ C, c ∩ c' != Null}
+        # It should be ready before initializing Cluster.
+        # Nothing but union of all the edge variables.
+        self.intersection_set_variables = set()
+
+        # The generation of all the Intersections of all
+        # the pairwise edges taken one at a time
+        for edge_pair in it.combinations([sorted(i) for i in model.edges()], 2):
+            self.intersection_set_variables.add(frozenset(edge_pair[0]) & frozenset(edge_pair[1]))
+
+        # Initialize each cluster for variables appearing in each of the edges.
+        # We also need the potentials of the current edges.
+        self.cluster_set = []
+        scope_list = [i.scope() for i in model.get_factors()]
+        for edge in [sorted(i) for i in model.edges()]:
+            index_of_factor = scope_list.index(list(edge))
+            self.cluster_set.append(self.Cluster(edge, self.intersection_set_variables,
+                                                 model.get_factors()[index_of_factor]))
+
+    class Cluster:
+
+        """
+        Inner class for representing a cluster.
+        A cluster is a subset of variables.
+
+        Parameters
+        ----------
+        set of variables: tuple
+                          This is the set of variables that form the cluster.
+
+        intersection_set_variables: set containing frozensets.
+                                    collection of intersection of all pairs of cluster variables.
+                                    For eg: {{C1 ∩ C2}, {C2 ∩ C3}, {C3 ∩ C1}} for clusters C1, C2 & C3.
+
+        cluster_potential: Factor
+                           Each cluster has a initial probability distribution provided beforehand.
+        """
+        def __init__(self, set_of_variables_for_cluster_c, intersection_set_variables, cluster_potential):
+            """
+            Initialization of the current cluster
+            """
+
+            # The variables with which the cluster is made of.
+            self.cluster_variables = set_of_variables_for_cluster_c
+
+            # The cluster potentials must be specified before only.
+            self.cluster_potential = copy.deepcopy(cluster_potential)
+
+            # Generate intersection sets for this cluster; S(c)
+            self.intersection_sets_for_cluster_c = [i.intersection(self.cluster_variables)
+                                                    for i in intersection_set_variables
+                                                    if i.intersection(self.cluster_variables)]
+
+            # Initialize messages from this cluster to its respective intersection sets
+            # \lambda_{c \rightarrow \s} = 1/|S(c)| * max_{x_{c-s}}{\theta_{ij}}
+            self.message_from_cluster = {}
+            for i in self.intersection_sets_for_cluster_c:
+                other_variables = list(set(set_of_variables_for_cluster_c)-i)
+                phi = copy.deepcopy(cluster_potential)
```

Comment:
@kislayabhi I think instead of using ```copy.deepcopy```, we can write a ```copy``` method in ```Factor``` class. This method would be returning a new factor using the same variables present in the cluster.

@ankurankan Your suggestions on this?